Atomic operations.

Apprentice-Alchemist opened the pull request:
The JIT implementation is based on what GCC emits for it's intrinsics.
The C implementation is copy pasted from what I did for hxcpp.

See https://github.com/HaxeFoundation/haxe/pull/10610

ncannasse commented:
I don't think this requires some specific opcodes, especially since the C version - which is supposed to be faster - is based on library function calls.

Apprentice-Alchemist commented:
Good point about library calls on MSVC, I should replace those with the intrinsics.

But for GCC/Clang it already uses instrinsics, which do not call library functions: [Godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYMQAdlIOAMngMmABy7gBGmMQSPgAOqAqEtgzObh7epHEJNgIBQaEsEVFcPpaY1klCBEzEBCnuniWYVtkMldUEuSHhkdEWVTV1aSX9HYFdBT3FAJQWqK7EyOwcAKQATADMgchuWADUy%2BuOCgToYaguAHQIB9jLGgCCd48beFS7WDRB6BCmAOLBckcpim%2By8jmWYPemE%2BmG%2BZgMRmBTzW6w%2BY12AAk/MZfo4gfcACoAeQAsgBJRxCZEbJqvKEwuEkoRAgBq2AASlNqajoeiscYWXjjITSRSqQ8UU0lLsAPQy3bBTDVWgAT12mAAbpEVQJMLtXEoFLtcY5SLtHAjgLsSLsmYKzcQ8MAEAQDgAxbmRYg23a%2Bv3%2BgOBoPBkOhsOB5YAVnBD39a1WisICEixrxZu2hitNrtjl2eCNEUCVoNsIuu2U9CY0rQgkdYVcBD1BGTu2ITU1hgIuyYRBYeGQeYYx0dQ/7Rv4xF2OvmuzQLBidEiF3jnoY%2BCoyNjgS7CFoxh7qD7yH36G%2BGpcPcXg67AComGbt7swiCIQAhbl0tFfCD8k3C4nkpSXKxq2mAEPMDC7KYB5HsYVBgcgCAnt895PmapgioBxhCNgACKxiUgSXLrO%2BEo0rQn68t%2B/KCviAFisB9y%2Bm24HEJBxhkoIkS0KgyAANawtgqiIZmmD3KeUCPueBg2PQN5TKhz4HKRzzrA4rzIl4AAim5MY%2Bu77r2/bGAorhhBA0mXvQ167HeD6CE%2BL5eCpKKUQyP7Yn%2BmEMU8zFgRBUGGYexnwQQiEmWZECKeh/6ikCOH4YRxEueRbljN8NFCt5QG%2BaBrHsZxjbEDx/GCcJCCieJ3wQFJF6yZg8moQAtEpJGruumk6RKW4OQZMHGYYZ51Vej52TZz6gilmxvF%2BsIeTiWX0TlIEsQF0FGceoXhYNUVmmEMXZdheEEUIRHKZ6FEzVRc2ZXRcVCIxfn5VBhXcbxAnoPca6SQ5ln1Y1e3Je1GkStpuk2X1G3GCQFnDdZo2oY%2BE1vh%2BV3ub%2Bi33Y9eVrUFsFbUhMPRYFh0JSdZ1tWRamXfS6XzbRsWAQ9uWrWxL1ccV72wkSxA/V2f2LgDjnnVT6kbqDXWPD1O57v1x6qDDAvww5Y1I05U1pdRnmY0z2Os%2BxeMhQhSEK7zxMYUtR2JadQOizTs0ZdiDPZczK3%2BWzHEcyVH0ABowzVv1ww1CmAyLqli514P6bLUNzjE1SYMYmDlaJsMySNKuIw5ycxGUjboKQuXhsX4aPm2MQGIsbCCOrqO01rC13bruWPknqi59YsL7OsWnqu3eewmHvrrcFx5xwnbciUYmC7fsqwAGxtx3%2Bdmms8/l5XmDVwQZpUGISiFyBJfhhbcVW%2BTB2W2TSVDzjHs5wP6Bh5K9vXY7Ao6z5bvPZ7RXe7Czh5wJyElPYAM8A78yDkLDeTAq6MB3n3ZesJbbhzXCDR4YNup6V6jHUek8KrTzTlZPUCN7JdhgXAmuk064O3ml5JarsmI2RYt3XuGhb4j1gsnUBM9UJrwoVveBq8F4sUvmfa%2BNtb76zys/VKaM6a3UZl/Jh0jf5vVKugEBBCwF812ErYOqEBHbxQZKNB4sMGS18tHQ2x4eJMCGunZWt4mC1ypprG62sm7KN9GXMCrDdjsMpkwzhxk7EoWEevMCYisISIpipJ6AUWKyOpu49%2BzsGF63dgVL2XN0B%2B15hAvRUCQ4BJMTSMxkcsEQ1wbBY4JAZ76JsqrX6YhXCYFcapVJdDP7LWCTYkyRA2yzzXhqVpmBokUnPjfIJCS2ajLcJgZJtJ5EN3SVjXKajOYaK0anQp%2BihbzLaSg2ZkFDmLKCaYjqEsOAzFoJwSMvBPAcC0KQVAnBcwKDmAsPUKIeCkAIJoG5Mw%2BIgHWOsC4YLIVQuhfPfQnBJCPMBa8zgvAFAgA0P8wFMw4CwCQHHRcZAKAQHxfQKIyAMxGC4OsDQXA%2BB0CKmiiAYQkVhECNUFUnA/lzm3kSBgqokVYBYJmcQzzeD4DbOULUaLRW%2BFUGUBsSw/nbiaEiiiYRiDsucFgJFBBHQsE5TcvgBhgAKBZHgTAAB3Ikucnl/P4IIEQYh2BSBkIIRQKh1Ayt0LSy0IBTDGHMGqtFkAZioBiC0aVTUiTrFRe2codgIAOEGJ4Wl/gxj5EKHoTIiQBDJqzfEHNDBOgZsmBYONLQ2gDBcPUPQpR42tBGMW7oRQ%2BjtDzbS447Qm0TCKDMT58xFh6F1ZgRVhq7kcAeaQJ5Ly3kcFUAADnnk1eekhdjAHJbsLgXALirF2BAU0uxcCEBtL8s0gCFykrnusLgUxeAAtFVMYFoLwXQtfZC2F46EVTqRbO1F6LMUPtIDixAIA5gEBiA2cglASWlthMelt9rhCiHEC6xD7q1A6rbKO0gFqNUxANbc%2B5iKZWzqJA2CDXZUBvAXUulds5LSbvBRoC4XA93noJVem9d6sVPrBRCt9r64UcC/dO3gv6LD/vvVoR9QnVjEZnSigD0mZhamIAkOwkggA%3D)

Simn commented:
What's the status here? I'm willing to merge the hxcpp PR, but I'm unsure where we're at for HL.

ncannasse commented:
That's a lot of code. I'm curious what's the rationale for supporting all 8/16/32/64 bits int values ? I think 32 + ptr only would be good enough.
Also, if the hl_atomic inlines are only meant to support HL/C, maybe they should be in hlc.h and not hl.h ?

Apprentice-Alchemist commented:
> That's a lot of code. I'm curious what's the rationale for supporting all 8/16/32/64 bits int values ? I think 32 + ptr only would be good enough.

Hashlink has 8/16/32/64 bit int types, so I thought it made sense to already implement atomic operations for all of them even though there are no AtomicI8/16/64 types (yet).

> Also, if the hl_atomic inlines are only meant to support HL/C, maybe they should be in hlc.h and not hl.h ?

Having them in hl.h could be useful if someone needs atomic operations when, for example, writing a hdll.

ncannasse commented:
I think we should keep things more simple with only int32 atomics and using primitives instead of opcodes.

Apprentice-Alchemist commented:
I've removed the 8/16/64 bit atomics, and switched everything to primitives.
I have kept the pointer atomics, it'd be a shame if AtomicObject were only implemented for Java.

Simn commented:
@ncannasse Should this be merged? I'm happy to handle the hxcpp and Haxe side of this, but we need a decision for HL first.

ncannasse commented:
I could merge the PR but it still contains some JIT support leftovers.

ncannasse commented:
Last change before merge : I think we should directly add atomics to threads.c ;)

ncannasse commented:
Thanks !
